backend/rates/coingecko.go

```diff
@@ -0,0 +1,153 @@
+
+		end := updater.histEarliestTimestamp("BTC")
+		if end.IsZero() {
+			end = time.Now()
+		}
+		// 90 days is the max interval which CoinGecko responds with hourly timeseries.
+		start := end.Add(-90 * 24 * time.Hour)
+		if time.Since(start) > 720*24*time.Hour {
+			// ~2 years of historical data is enough for now.
+			// TODO: Where do we stop?
+			return
+		}
+
+		err := updater.updateHist("BTC", "USD", start, end)
+		if err != nil {
+			updater.log.Printf("updateHist(%s, %s): %v", start, end, err)
+			untilNext = time.Minute // TODO: exponential backoff
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+func (updater *RateUpdater) updateHist(coin, fiat string, start, end time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	rates, err := updater.fetchMarketRange(ctx, coin, fiat, start, end)
+	if err != nil {
+		return err
+	}
+
+	updater.histMu.Lock()
+	defer updater.histMu.Unlock()
+	// TODO: updater.hist[coin] and rates may have duplicate timestamps.
+	a := append(updater.hist[coin], rates...)
+	sort.Slice(updater.hist[coin], func(i, j int) bool {
+		return a[i].timestamp.Before(a[j].timestamp)
+	})
+	updater.hist[coin] = a
+
+	updater.Notify(observable.Event{
+		Subject: "historical_rates",
+		Action:  action.Reload,
+	})
+	return nil
+}
+
+func (updater *RateUpdater) histLatestTimestamp(coin string) time.Time {
+	updater.histMu.RLock()
+	defer updater.histMu.RUnlock()
+	var t time.Time
+	if n := len(updater.hist[coin]); n > 0 {
+		t = updater.hist[coin][n-1].timestamp
+	}
+	return t
+}
+
+func (updater *RateUpdater) histEarliestTimestamp(coin string) time.Time {
+	updater.histMu.RLock()
+	defer updater.histMu.RUnlock()
+	var t time.Time
+	if len(updater.hist[coin]) > 0 {
+		t = updater.hist[coin][0].timestamp
+	}
+	return t
+}
+
+func (updater *RateUpdater) fetchMarketRange(ctx context.Context, coin, fiat string, start, end time.Time) ([]exchangeRate, error) {
+	param := url.Values{
+		"from":        {strconv.FormatInt(start.Unix(), 10)},
+		"to":          {strconv.FormatInt(end.Unix(), 10)},
+		"vs_currency": {fiat},
+	}
+	endpoint := fmt.Sprintf("%s/coins/%s/market_chart/range?%s", updater.coingeckoURL, coin, param.Encode())
+	req, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := updater.httpClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close() //nolint:errcheck
+	var jsonBody struct {
+		Prices [][2]float64 // [timestamp, value], sorted by timestamp asc
+	}
+	// 1Mb is more than enough for a single response.
+	// For comparison, a range of 15 days is about 14Kb.
+	if err := json.NewDecoder(io.LimitReader(res.Body, 1<<20)).Decode(&jsonBody); err != nil {
+		return nil, err
+	}
+
+	rates := make([]exchangeRate, len(jsonBody.Prices))
+	for i, v := range jsonBody.Prices {
+		rates[i] = exchangeRate{
+			timestamp: time.Unix(0, int64(v[0])), // local timezone
```

Comment:
this does not work at all, i get only 1970- timestamps. the arguments seem reversed, but when i reverse them i get timestamps in the year 52452 :joy:

Reply:
that should work. the args are correct: https://golang.org/pkg/time/#Unix

> ... It is valid to pass nsec outside the range [0, 999999999].

you can see the source code: https://golang.org/src/time/time.go?s=39974:40011#L1282

> need to reverse args and divide by 1000.

right, or don't reverse the args and multiple by 1000. I forgot the 2nd arg is nanoseconds.

Reply:
but yeah, either way there's a mistake. doesn't work as is :sweat_smile:

Reply:
done